### PR TITLE
Kconfig: Expose Gcoap configurations

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description
This PR adds the `CONFIG_` prefix to Gcoap's configuration macros and then exposes those options to Kconfig.

The approach is to have a menuconfig entry for the module (`KCONFIG_MODULE_GCOAP`) which enables/disables the configuration of Gcoap via Kconfig. This entry will only show when the module is being used (`depends on MODULE_GCOAP`).

`MODULE_` symbols are declared in the `Kconfig.dep` file generated from `USEMODULES`:
https://github.com/RIOT-OS/RIOT/blob/c6b9f9509c9ec1608489a9718380d3bbabb5fefe/makefiles/kconfig.mk#L54-L58

### Testing procedure
- Gcoap's example application should still be working as usual (`make all`). In this case the applied configurations should be the ones declared in `gcoap.h`.
- Running `make menuconfig` in Gcoap's example application should reveal the menuconfig interface and Gcoap's configurations. Configurations selected via this interface should be applied to the application.

### Issues/PRs references
#12888

### CODE REVIEW

File: Kconfig [hunk 1]
```diff
@@ -9,6 +9,8 @@ mainmenu "RIOT Configuration"
 # For now, get used modules as macros from this file (see kconfig.mk)
 osource "$(KCONFIG_GENERATED_DEPENDENCIES)"
 
+rsource "sys/Kconfig"
```
Comment (cgundogan) [inline Kconfig]: is that supposed to be `osource`?

Comment (leandrolanzieri) [reply Kconfig]: Not in this case. `osource` is for 'optional includes', so the pattern may not match any file. In this case `rsource` is for including files using a relative path.

All the options are documented [here](https://github.com/ulfalizer/Kconfiglib#kconfig-extensions).

**Comment by leandrolanzieri** [comment]
Rebased to master, 'system' and 'net' Kconfig files are merged now.

**Comment by tcschmidt** [comment]
ping @leandrolanzieri @cgundogan

**Comment by leandrolanzieri** [comment]
> On a side note: I really like this interface. It is very simple to discover available configurations :) Much better than searching for the correct header file or doxygen page, IMO ..

:+1: 

> Any ideas what might be wrong?

Yes, as this was one of the first PRs we still didn't have the Kconfig symbols available in Makefiles, so it is expected to have errors if configurations are set via CFLAGS. I will update this to check for Kconfig usage and avoid conflicts.

File: sys/net/application_layer/gcoap/Kconfig [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+config GCOAP_OBS_CLIENTS_MAX
+    int "Maximum number of Observe clients"
+    default 2
+
+config GCOAP_OBS_REGISTRATIONS_MAX
+    int "Maximum number of registrations for Observable resources"
+    default 2
+
+config GCOAP_OBS_VALUE_WIDTH
```
Comment (leandrolanzieri) [inline sys/net/application_layer/gcoap/Kconfig]: @kb2ma looking at https://github.com/RIOT-OS/RIOT/blob/f6d750afe9420761dd3fa13a5791a35ca4a046f5/sys/include/net/gcoap.h#L581-L587

Is the range of this value [1 - 3]?

**Comment by leandrolanzieri** [comment]
@cgundogan I rebased so I can use the change of #12913

File: sys/net/application_layer/gcoap/Kconfig (hunk 2 above)
Comment (leandrolanzieri) [reply sys/net/application_layer/gcoap/Kconfig]: As it is, the option must take one of these values, so I added the range in ad75697d891699f19a390e3ff83e626a3b0292ef.

**Comment by leandrolanzieri** [comment]
> Tested the fixup commit: observable tick length ranges now from 1..3 and kconfig does not allow values outside this range.
> 
> ACK. Please squash!

@cgundogan squashed

File: sys/net/application_layer/gcoap/Kconfig [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        allow a client to detect message reordering withing the network latency
+        period (128 sec). For resources that change slowly, the reduced
+        message length is useful when packet size is limited.
+
+endmenu # Observe options
+
+menu "Timeouts and retries"
+
+config GCOAP_RECV_TIMEOUT
+    int "Incoming message timeout"
```
Comment (kaspar030) [inline sys/net/application_layer/gcoap/Kconfig]: general question on KConfig: are these converted to explicitly typed literals (`1000000LU`)?

Comment (leandrolanzieri) [reply sys/net/application_layer/gcoap/Kconfig]: No, Kconfig does not add suffixes to these literals, the header file will just contain `1000000`. If needed, could we add a cast when using the macro?

Comment (miri64) [reply sys/net/application_layer/gcoap/Kconfig]: So all type checking by the compiler is removed?

Comment (cgundogan) [reply sys/net/application_layer/gcoap/Kconfig]: >So all type checking by the compiler is removed?

no, the compiler checks the type. What is removed is the possibility to set types for integer literals. This only affects CONFIG symbols that are used in arithmetic expressions. Also, the compiler still throws an error if integer widths do not match. Using an explicit cast (...) solves this problem.

Comment (cgundogan) [reply sys/net/application_layer/gcoap/Kconfig]: @kaspar030 ready to merge?

Comment (cgundogan) [reply sys/net/application_layer/gcoap/Kconfig]: Murdock shows green lights. Let's merge this now, but we can still continue the discussion in this thread for this particular kconfig property, if needed.

**Pull request merged**